sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala

```diff
@@ -324,6 +324,8 @@ private[sql] class DynamicPartitionWriterContainer(
 
   private def bucketIdExpression: Option[Expression] = for {
     BucketSpec(numBuckets, _, _) <- bucketSpec
+  // Use `Pmod` to match the behaviour of `Utils.nonNegativeMod`, which is the function used to
```

Comment:
`Pmod` works for all numerics but `Utils.nonNegativeMod` only works for int.
How about we also use `Pmod` in shuffle so that we don't need to depend on `Utils.nonNegativeMod`?